src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -236,14 +243,26 @@ SafeToPushdownUnionSubquery(RelationRestrictionContext *restrictionContext)
 		}
 		else if (unionQueryPartitionKeyIndex != partitionKeyIndex)
 		{
-			return false;
+			continue;
 		}
 
 		AddToAttributeEquivalenceClass(&attributeEquivalance, relationPlannerRoot,
 									   varToBeAdded);
 	}
 
-	return EquivalenceListContainsRelationsEquality(list_make1(attributeEquivalance),
+	relationRestrictionAttributeEquivalenceList =
```

Comment:
Given that these code paths are slightly complex to the readers, could you add a comment here? I think a variation of the PR description should be sufficient.